sourcers/json/json.go

```diff
@@ -0,0 +1,62 @@
+package json
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+
+	"github.com/ardanlabs/conf"
+)
+
+type JSONSourcer struct {
+	m map[string]string
+}
+
+// NewSource returns a conf.Sourcer and, potentially, an error if a
+// read error occurs or the Reader contains an invalid JSON document.
+func NewSource(r io.Reader) (conf.Sourcer, error) {
+	if r == nil {
```

Comment:
I am not a fan of this kind of code. It adds more code than is required and is the caller doesn't make the call properly, let the application panic. That's better because this is a serious situation.

Reply:
Understood.

A conundrum arises, though:

Say a developer wants to support an optional `json` config.  Without the `nii` reader check, `conf` forces the developer to check for the config file's existence prior to using `conf.Parse`.

Something I also just realized: with the`sources ...Sourcer`-style `Parse`, the final `Sourcer` would take precedence if a user had both, say, `json` and `yaml` config files before being overwritten by the environment and flags.  How do you want to handle that?

Reply:
That is the callers concern not mine. You should never make the API call with nil.

I have to look at the code but are you saying these sourcers take precedence over the env and command line flags?

Reply:
env and cli take precedence over the configs.

That still leaves the `...Sourcer` issue, though.

`conf.Parse(args, "TEST", &cfg, jsonFile, yamlFile)` parses the json, then yaml, then env, and finally the flags.  That puts the onus on the caller to check for the right file first.  (Or db/kv store/whatever.)

Reply:
I suppose the confusion lies in `conf.Parse`'s varargs `sourcers` parameter.  If there should be at most one extra `Sourcer`, why not create a separate `conf.ParseWithSourcer` func?

With the varargs parameter, `conf.Parse` currently loops through all of the possibilities, then overwrites as needed with env/flags.

See [conf.go#L113-L133](https://github.com/ardanlabs/conf/blob/master/conf.go#L113-L133).